### PR TITLE
DrivAerML: EMA=0.9995 + gc=0.5 + 16 attention heads

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current DrivAerML champion (PR #3072) uses 8 attention heads at 64d/head (4L/512d/8H config). We hypothesize that increasing to 16 heads (32d/head, same total hidden dim 512d) will enable finer-grained spatial attention — each head specialises on a narrower subspace of the surface geometry, which may improve local pressure feature resolution on aerodynamic surfaces.

Prior work: senku tested 4H and 16H variants in PR #3127 **without** EMA+gc and did not beat baseline. That result is insufficient to conclude heads don't matter — EMA=0.9995 + gc=0.5 is a critical stability mechanism on DrivAerML (EMA alone diverges at ep28 without gc). This PR adds EMA+gc to the 16H variant to give it a fair test.

Reference: Multi-head attention head count vs. head dimension tradeoffs are well-studied in NLP (Voita et al. 2019 "Analyzing Multi-Head Self-Attention") and have shown that head specialization is architecture-dependent.

## Instructions

Start from the DrivAerML champion config (4L/512d/8H, EMA=0.9995, gc=0.5) and change only the number of attention heads from 8 to 16:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 16 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --wandb-group dm-ema-heads
```

Key change vs. champion: `--model-heads 16` (was 8). Head dim goes from 64d to 32d. All other hyperparameters are identical to #3072.

**What to report:**
- Best `val_primary/surface_rel_l2_pct` and the epoch it was achieved
- Training stability (any divergence or instability events?)
- VRAM usage comparison vs. 8H champion
- W&B run link and run ID

## Baseline

Current DrivAerML best (PR #3072, eren — EMA=0.9995 + gc=0.5, 4L/512d/8H):

| Metric | Value |
|--------|-------|
| `val_primary/surface_rel_l2_pct` | **3.833%** |
| Epoch | 511 |

- Baseline W&B run: `ncl1dh88` (eren/dm-ema-9995-gc05)
- External target (AB-UPT): <3.71% (~500 epochs) — gap is 0.123pp

Reproduce champion:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT griffith:

## Results

### DrivAerML: 16 attention heads (32d/head) vs. champion 8 heads (64d/head)

| Metric | 16H (this run) | 8H Champion (#3072) | Delta |
|--------|---------------|---------------------|-------|
| `val_primary/surface_rel_l2_pct` | **5.358%** (ep156) | **3.833%** (ep511) | +1.525 pp (worse) |
| Stable training epochs | 158 (diverged ep159) | 511+ (no divergence) | -353 epochs |
| External target (AB-UPT) | >3.71% | <3.71% | — |

**Verdict: NEGATIVE.** 16 heads does not beat the 8-head champion. The 16H variant diverges at epoch 159, preventing it from reaching the deep basins the champion finds at epoch 400+.

### Training stability

- **Epochs 1-156:** Perfect run — 155/156 new bests, very smooth descent. EMA warmed up by ep46. Grad norms decreased from ~10 → ~0.27 over training.
- **Epoch 159:** Gradient explosion — grad norm jumped from 0.39 → 34.7 at LR=4.5e-4 (cosine near-peak). gc=0.5 clipped but couldn't contain the cascade.
- **Epochs 160-162:** Irreversible divergence. Grad norms: 55.8 → 238.9 → 268.4. Val: 14.9% → 22.5% → 30.7%. Training killed at ep162.
- **Divergence cause:** The 16H variant with 32d/head appears to have less stable attention dynamics than 8H/64d/head. The smaller per-head subspace may be more susceptible to gradient explosions at cosine LR peaks. The champion (8H/64d) ran 511 epochs without any instability.

### VRAM usage

- **16H (this run):** ~7.0 GB peak on RTX PRO 6000 (97.9 GB available)
- **8H champion:** Similar — same total hidden dim (512d), VRAM is dominated by data loading not attention heads

### W&B

- **Run ID:** `s6h3y0w0`
- **Run link:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/s6h3y0w0
- **Group:** `dm-ema-heads`

### Command used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 16 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --wandb-group dm-ema-heads \
  --agent griffith --wandb-name "griffith/dm-ema-16h-heads-v3"
```

### What happened

The 16-head variant trained beautifully for 156 epochs with perfect stability, then catastrophically diverged at epoch 159. The gradient explosion cascaded over 4 epochs (grad norms: 0.39 → 34.7 → 55.8 → 238.9 → 268.4) despite gc=0.5. The EMA model preserved the best checkpoint at 5.358%, but this is 1.525 pp worse than the 8-head champion's 3.833%.

The root cause is that the 8-head champion reaches its best at epoch 511 — it needs ~350 more epochs of stable training beyond where the 16-head variant diverges. Even if 16H were slightly faster to converge per-epoch (which the data doesn't strongly support), the early divergence prevents it from reaching the champion's depth.

This result is consistent with senku's earlier finding in PR #3127 (16H without EMA+gc also didn't beat baseline). Adding EMA+gc gave it a fair test but the conclusion holds: 8 heads at 64d/head is the better configuration for DrivAerML.

### Suggested follow-ups

- **12 heads (42-43d/head):** A middle ground between 8H and 16H might preserve more stability while still gaining some multi-head specialization benefit.
- **16H + lower LR:** The divergence occurs at LR peaks. A lower peak LR (e.g., 3e-4 instead of 5e-4) or a longer cosine cycle (T_max=50) might stabilize the 16H variant long enough to test whether it can match the champion.
- **16H + gc=0.3:** Tighter gradient clipping might prevent the specific instability mode. The champion uses gc=0.5 but 16H may need tighter control.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3160 because ADVISOR: Closing — 16H (32d/head) diverges at ep159, too early to compete. Best: 5.358% val vs baseline 3.833%. The champion (8H/64d) needs 511 epochs to reach 3.833% — 16H can't survive that long due to attention instability at cosine LR peaks. 8H/64d confirmed as the right head configuration for DM. GPU reassigned..
